### ISSUE TITLE
Java adl constructors should take List<> (or Collection<>), not ArrayList<>

### ISSUE THREAD

**Issue by latos** [timbod7/adl#49]

**Comment by timbod7** [comment]
Currently `ArrayList<>` is used consistently, and constructors just copy arrays references and not the contents. We'd need to either work out an optimal alternative to this approach, or provide explicit control through annotations.

**Comment by timbod7** [comment]
Currently the ADL java backend has a 1 to 1 mapping between ADL type and java types.

But I agree it  would be more idiomatic to have 1 ADL type map to 2 java types, one for construction (ie ArrayList), and 1 for interfaces (ie List).

Similarly for  `HashMap/Map` and `HashSet/Set`.

**Comment by timbod7** [comment]
A sometimes useful workaround if you are using the java streaming API is:

```
 ArrayList<T> result = XXXX.stream()
    .map(...)
    .collect(Collectors.toCollection(ArrayList::new));
```

**Comment by timbod7** [comment]
Fixed with: e4bfb762b7d6ec6928f4e92793dd5cc5afdbe787
